usaspending_api/broker/management/commands/load_fpds_from_broker.py

```diff
@@ -0,0 +1,135 @@
+
+    @staticmethod
+    def next_file_batch_generator(file):
+        while True:
+            lines = file.readlines(CHUNK_SIZE)
+            lines = [line.decode("utf-8") for line in lines]
+            if len(lines) == 0:
+                break
+            yield lines
+
+    def load_fpds_from_file(self, file_path):
+        with RetrieveFileFromUri(file_path).get_file_object() as file:
+            for next_batch in self.next_file_batch_generator(file):
+                id_list = [int(re.search(r"\d+", x).group()) for x in next_batch]
+                logger.info(
+                    "Loading next batch from provided file (size: {}, ids {}-{})...".format(
+                        len(id_list), id_list[0], id_list[-1]
+                    )
+                )
+                self.modified_award_ids.extend(run_fpds_load(id_list))
+
+    def add_arguments(self, parser):
+        mutually_exclusive_group = parser.add_mutually_exclusive_group()
+
+        mutually_exclusive_group.add_argument(
+            "--ids",
+            nargs="+",
+            type=int,
+            help="(OPTIONAL) detached_award_procurement_ids of FPDS transactions to load/reload from Broker",
+        )
+        mutually_exclusive_group.add_argument(
+            "--date",
+            dest="date",
+            type=datetime_command_line_argument_type(naive=True),  # Broker date/times are naive.
+            help="Load or Reload all FPDS records from the provided date to the current time. YYYY-MM-DD format",
+        )
+        mutually_exclusive_group.add_argument(
+            "--since-last-load",
+            action="store_true",
+            help="Equivalent to loading from date, but date is drawn from last update date recorded in DB",
+        )
+        mutually_exclusive_group.add_argument(
+            "--file",
+            metavar="FILEPATH",
+            type=str,
+            help="A file containing only transaction IDs (detached_award_procurement_id) "
+            "to reload, one ID per line. Nonexistent IDs will be ignored.",
+        )
+        mutually_exclusive_group.add_argument(
+            "--reload-all",
+            action="store_true",
+            help="Script will load or reload all FPDS records in broker database, from all time",
+        )
+
+    def handle(self, *args, **options):
+        # loads can take a while, so we record last updated date from the start of all transactions
+        last_update_time = datetime.now()
+
+        if options["reload_all"]:
+            self.load_fpds_from_date(None)
+
+        if options["ids"]:
+            self.modified_award_ids.extend(run_fpds_load(options["ids"]))
+
+        if options["file"]:
+            self.load_fpds_from_file(options["file"])
+
+        if options["date"]:
+            self.load_fpds_from_date(options["date"])
+
+        if options["since_last_load"]:
+            self.load_fpds_from_date(get_last_load_date("fpds"))
```

Comment:
Tragically I was able to confirm this with broker. Are we sure we don't have enough confidence to cut down this interval? Right now this is doubling our run time.

Reply:
We could make the algorithm that checks for new records smarter.  Right now we're just checking by date, but what if we checked by id and date and only processed records in the last 24~48 hours that either are missing from USAS based on id or have a newer update date.

It would take a little longer to figure out, but it would cut down on processing which is the real killer.

Reply:
@BrianZito What is doubling the runtime? If I understand `get_last_load_date()` correctly, it will return the stored datetime from `external_data_load_date` when `lookback_minutes` = None. That datetime is the start datetime from the previous successful execution.

The default value is returned only if it was provided to the function and the previous script datetime was unable to be retrieved from the DB.

The default datetime is set to -1 day but hopefully the script will never use that value.

Reply:
@kbard isn't the issue you described pertinent to only FABS transactions?

Reply:
@tony-sappe, of course you're right.  So I guess my original claim was incorrect.  Digging back through the history, it seems that logic has been in the file long before I even started on  this project.  I guess I confused it with the FABS logic.  Although I have to wonder if/why FPDS is immune to the race conditions we found in FABS... but that's a different investigation ticket.

Sorry @BrianZito.  Please revert the change.

**HOWEVER**, after looking again at the new code, I notice that this bit won't function if there's no date in the database.  Please provide a default to `get_last_load_date`.

Reply:
I did some testing, and with a default of None it simply reloads all transactions, which I would argue is the correct functionality. I'll see what I can do to put some automated tests on that.

Reply:
Any response to the last comment?

Reply:
Sorry, must have gotten this lost in the confusion. I suppose this comes down to a subjective case, that would need UX experience to figure out authoritatively. Given that I don't expect this case to ever occur in the lifetime of this code, I'm more than happy to just have it crash and if anyone really wants it to do something else they can always make another PR.

Reply:
Just looked at the recent code, I didn't notice the new conditional which will work nicely. A helpful message would be appreciated

```python
raise ValueError("No last load date for FPDS stored in the database")
```